sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/export/IntervalMetricReader.java

```diff
@@ -141,8 +140,9 @@ CompletableResultCode doRun() {
                 flushResult.succeed();
                 exportAvailable.set(true);
               });
-        } catch (RuntimeException e) {
-          logger.log(Level.WARNING, "Exporter threw an Exception", e);
+        } catch (Throwable t) {
```

Comment:
Exception should be sufficient but we are anyway not really consistent here

Reply:
I don't think Exception is sufficient. If a particularly large "batch" of metrics causes an OutOfMemoryError to be thrown for example, this should not permanently break the IntervalMetricReader.

Reply:
Alright, if this would happen at once before the memory for the entire batch is allocated, it might recover from that situation.

Reply:
It sounds like it's time for us to follow the tradition of all the Java libraries out there and add a throwIfFatal helper (originally from Scala)

https://github.com/line/armeria/blob/d321dda1879ff4e24df4a3a55d620c1d86773146/core/src/main/java/com/linecorp/armeria/common/util/Exceptions.java#L258

We don't necessarily need to do it in this PR but we do need to do it eventually.

Reply:
That pattern does fail on OutOfMemoryError btw - it should I think, if metrics have caused an app to go over memory they will continue to do so most likely, turning ourselves off is probably better (ideally we shutdown the whole SDK in that case).

Reply:
> That pattern does fail on OutOfMemoryError btw

Not sure what you mean? The linked function does rethrow VirtualMachineError which is the base of OOMError

Reply:
Right sorry by "fail" I meant crashes the app. I think it's generally ok as a general principle to crash on OOM since the chance of recoverability is so slim.

Reply:
But we shouldn't crash on the silly `AssertionError` or user exceptions that happen to extend `Error` mistakenly.

Reply:
>  I think it's generally ok as a general principle to crash on OOM since the chance of recoverability is so slim.

That depends on where the OOM comes from. If it comes from something like `new MyHarmlessObject()`, the chance of recovery is slim. If it comes from `new byte[1024 * 1024 * 1024]`, then things are different.

> But we shouldn't crash on the silly AssertionError

I think AssertionErrors can be seen as pretty fatal: They basically mean that the code detected a programming error.

However, I think the main issue here is that we would not crash either way. If the exception flies out of our `run`,  it just silently swallowed (would be returned from Future.get(), but we don't keep a reference to that around).

Reply:
> If it comes from new byte[1024 * 1024 * 1024], then things are different.

Yeah the notion though is that if this happens once, in all likelihood it's happening again so best to shutdown.

But yeah let's address it separately, not throwing on normal exceptions is more important.